### PR TITLE
Implementing a few switch statements

### PR DESCRIPTION
**Pull request by nate-thegrate**
I really like how patterns can be used for variable assignment and avoiding duplicated logic. (related: #150942)

```dart
// before
final GestureRecognizer? recognizer = info.recognizer;
if (recognizer is TapGestureRecognizer) {
  if (recognizer.onTap != null) {
    configuration.onTap = recognizer.onTap;
    configuration.isLink = true;
  }
} else if (recognizer is DoubleTapGestureRecognizer) {
  if (recognizer.onDoubleTap != null) {
    configuration.onTap = recognizer.onDoubleTap;
    configuration.isLink = true;
  }
}


// after
switch (info.recognizer) {
  case TapGestureRecognizer(onTap: final VoidCallback? onTap):
  case DoubleTapGestureRecognizer(onDoubleTap: final VoidCallback? onTap):
    if (onTap != null) {
      configuration.onTap = onTap;
      configuration.isLink = true;
    }
}
```

### CODE REVIEW

**Review by victorsanni — COMMENTED** [review]
LGTM logic wise. 

Style-wise, I personally think `Class(:parameter)`, while less verbose, can be a little disorienting to read than just `Class(parameter: parameter)`.  I think it's in the same spirit as the section of the style guide to [avoid anonymous parameter names](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-anonymous-parameter-names).

**Comment by victorsanni** [comment]
It sort of reminds me of how Python has quirky inbuilt tips and tricks that serve to reduce verbosity but make it much harder to read and non-obvious to a newcomer.

File: dev/tools/localization/bin/gen_date_localizations.dart [hunk 1]
```diff
@@ -164,61 +164,51 @@ String _jsonToMapEntry(String key, dynamic value) {
...
+      final Type type = json.first.runtimeType;
+      final StringBuffer buffer = StringBuffer('const <$type>[');
+      for (final dynamic value in json) {
+        buffer.writeln('${_jsonToMap(value)},');
+      }
+      buffer.write(']');
+      return buffer.toString();
+    case Map<String, dynamic>():
+      final StringBuffer buffer = StringBuffer('<String, Object>{');
+      for (final MapEntry<String, dynamic>(:String key, :dynamic value) in json.entries) {
```
[the remark this answers is not in the document]
Comment (nate-thegrate) [reply dev/tools/localization/bin/gen_date_localizations.dart:182]: [`avoid_function_literals_in_foreach_calls`](https://dart.dev/tools/linter-rules/avoid_function_literals_in_foreach_calls) is one of Dart/Flutter's recommended linter rules.

However, it's not mentioned in the Flutter repo style guide, and you make a good point about readability. (Not to mention the fact that I wasn't consistent between this function and the one below 😅)

I opened https://github.com/dart-lang/sdk/issues/55709 a couple of months ago, and if/when that's resolved we won't need duplicate type annotations.

Until then, I agree that using `forEach` is the best way to go.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 2]
```diff
@@ -1452,25 +1452,21 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
           ..sortKey = OrdinalSortKey(ordinal++)
           ..textDirection = initialDirection
           ..attributedLabel = AttributedString(info.semanticsLabel ?? info.text, attributes: info.stringAttributes);
-        final GestureRecognizer? recognizer = info.recognizer;
-        if (recognizer != null) {
-          if (recognizer is TapGestureRecognizer) {
-            if (recognizer.onTap != null) {
-              configuration.onTap = recognizer.onTap;
+        switch (info.recognizer) {
+          case TapGestureRecognizer(:final VoidCallback? onTap):
```
[the remark this answers is not in the document]
Comment (nate-thegrate) [reply packages/flutter/lib/src/rendering/editable.dart:1456]: I have somewhat mixed feelings about this—the [avoid anonymous parameter names](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-anonymous-parameter-names) is about removing a name entirely, whereas in this situation, it's removing a duplicate name from the same line.

But with regard to this function, adding `onTap: ` would make it parallel to the case right below it, which I agree is an improvement for readability. I'll definitely update this one.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 3]
```diff
@@ -3799,27 +3799,17 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
-    if (!_scrollableNotificationIsFromSameSubtree(notification.context)) {
-      return;
+    switch (notification) {
+      case ScrollStartNotification() when _dataWhenToolbarShowScheduled != null:
+      case ScrollEndNotification() when _dataWhenToolbarShowScheduled == null:
+        break;
+      case ScrollEndNotification() when _dataWhenToolbarShowScheduled!.value != _value:
+        _dataWhenToolbarShowScheduled = null;
+        _disposeScrollNotificationObserver();
+      case ScrollNotification(:final BuildContext? context)
```
Comment (victorsanni) [inline packages/flutter/lib/src/widgets/editable_text.dart:3809]: Here, for instance, I missed the colon on first glance and thought it was just `case ScrollNotification(final BuildContext? context)` (doesn't help that `:` and `final` have the same color highlight). It took a minute to read carefully again before I understood what was going on.

**Review by victorsanni — COMMENTED** [review]
I can imagine someone new to Dart seeing it and thinking "Dart is one of those obscure quirky languages and I don't understand what's going on".

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 3 above)
Comment (victorsanni) [reply packages/flutter/lib/src/widgets/editable_text.dart:3809]: But maybe it's just me. I am new to Dart myself :)

Comment (nate-thegrate) [reply packages/flutter/lib/src/widgets/editable_text.dart:3809]: In this case (no pun intended), I'd prefer keeping it as-is.

<br>

The Dart documentation shows this concept in the [patterns overview](https://dart.dev/language/patterns#for-and-for-in-loops):

> …This allows you to simplify the variable pattern from something redundant like `key: key` to just `:key`

(and it goes into a bit more detail in the [pattern types](https://dart.dev/language/pattern-types#object) page)

**Comment by nate-thegrate** [comment]
> doesn't help that `:` and `final` have the same color highlight

Yes, I can identify with that. Same thing goes with enum values: they have a distinct color in VS Code, but on GitHub they look just like other static class members.

**Review by victorsanni — APPROVED** [review]

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 3 above)
Comment (victorsanni) [reply packages/flutter/lib/src/widgets/editable_text.dart:3809]: SGTM.

**Comment by nate-thegrate** [comment]
Thank you Victor!

I think we're good to merge, since test updates are included in this PR.
(My plan was to ping `@test-exemption-reviewer` anyway, but it looks like they're pretty backed up as of now.)

**Pull request merged**